### DISCUSSION TITLE
v8 and Apexcharts-card

### DISCUSSION THREAD

**Discussion by gorstj** [BottlecapDave/HomeAssistant-OctopusEnergy#354] [General]
With change to version 8 the attributes for the future rates have changed structure.

I have had to change my apexcharts-card data generator statement (which I copy&pasted from somewhere!.

Here is a data_generator statement which works for me:

```
    data_generator: |
      return entity.attributes.all_rates.map((entry) => {
          return [new Date(entry.valid_from), entry.value_inc_vat];
        });
```

(I think rates-->all_rates_map, valid_from and value_inc_vat have all changed)

**Comment by gorstj** [comment]
Odd situation that the current rate is in GBP (£) but the attribute future rates are in pence....

**Reply by BottlecapDave** [reply]
I've updated the docs to relfect your discovered changes. The values are different representations (and always have been) because pence is not a valid monetary value in HA sensors (which change how data is visualised) but the attributes represent the raw value from OE (which is reported in pence)

**Comment by simonpjcox** [comment]
Thanks - I was wondering why the Import and Export Rates chart from the community page was no longer working - using the above seems to have fixed it (though I used entry.value_inc_vat/100 to get it formatted properly for pence). On a related note when I was first setting up that chart the other week I noticed the import price in the legend wasn't reflecting the current price when using the code on the Community page.  I had to add the following to get it to show properly (sorry can't seem to get the code to format properly!):

>     - entity: sensor.octopus_energy_electricity_[[SERIAL]]_[[MPAN]]_current_rate
      name: Import
      show:
        in_header: before_now
      curve: stepline
      data_generator: |
        return entity.attributes.rates.map((entry) => {
          return [new Date(entry.from), entry.rate/100];
        });

**Comment by StephenMilner** [comment]
I'm using a Tileboard rates customisation and I think I'm suffering from the same problem - Hopefully someone can point out the changes I need to make to get this back up and running.

I start the tile in the normal way:

title: 'Octopus Rates',
position: [0, 2.5],
type: TYPES.TEXT_LIST,
classes: ['-agile-rates'],
state: false,
width: 1.5,
height: 2,
id: {},

I then want to display 12 rates so I start with this:

list: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12].map(function (j) {
    var entity = 'sensor.octopus_energy_electricity_X_Y_current_rate';
    var getItem = function(ctx, index) {                                                                 
        return ctx.states[entity].attributes.rates[index]
    }; 

[...]

}

Is my entity definition now incorrect?

I refer to individual rates like this:

var rate = getItem(this, slot).rate;

and dates/times like this:

var fromdatetime = getItem(this, slot).from; 

Slot is calculated to the be the current rate plus the addition  of j, as defined above so the next 12 rates are displayed.

Any help greatly appreciated.

Apologies for the formatting - Guthub doesn't seem to like OSX or my mac.

**Comment by StephenMilner** [comment]
For information I've tried this:

var getItem = function(ctx, index) {                                                                 
    return ctx.states[entity].attributes.all_rates.map[index]                                          
 };

And to get a rate, this:

var rate = getItem(this, slot).value_inc_vat; 

Without any joy

**Comment by StephenMilner** [comment]
For reference for others - I didn't need the map in my syntax.

This resolved the problem:

var getItem = function(ctx, index) {
return ctx.states[entity].attributes.all_rates.[index]
};

**Comment by StephenMilner** [comment]
What I did notice is that before 4pm my rates seemed to be 24 hours old, and after the 4pm update my calculation for slot was a few hours early.

I will need to do some more investigations to calculate the slot correctly.

**Comment by mdeakin99** [comment]
Interested to understand the recommended end solution/card to use to get the rates/times displayed correctly
